services/web-server/src/servers/oauth2.js

```diff
@@ -0,0 +1,306 @@
+    }
+
+    const currentUser = await strategies[user.identityProviderId].userFromIdentity(user.identity);
+
+    await AuthorizationCode.create({
+      code,
+      // OAuth2 client
+      clientId: client.clientId,
+      redirectUri: redirectURI,
+      identity: user.identity,
+      identityProviderId: user.identityProviderId,
+      // The access token we give to third parties
+      accessToken: new Buffer.from(taskcluster.slugid()).toString('base64'),
+      // A maximum of 10 minutes is recommended in https://tools.ietf.org/html/rfc6749#section-4.1.2
+      expires: taskcluster.fromNow('10 minutes'),
+      clientDetails: {
+        clientId: `${user.identity}/${client.clientId}-${taskcluster.slugid().slice(0, 6)}`,
+        description: `Client generated by ${user.identity} for OAuth2 Client ${client.clientId}`,
+        scopes: scopeIntersection(ares.scope, currentUser.scopes()),
+        expires: ares.expires ?
+          ares.expires > taskcluster.fromNow(client.maxExpires) ?
+            taskcluster.fromNow(client.maxExpires).toISOString() :
+            ares.expires.toISOString()
+          : taskcluster.fromNow(client.maxExpires).toISOString(),
+        deleteOnExpiration: true,
+      },
+    }, true);
+
+    return done(null, code);
+  }));
+
+  /**
+   * After a client has obtained an authorization grant from the user,
+   * we exchange the authorization code for an access token.
+   *
+   * The callback accepts the `client`, which is exchanging `code` and any
+   * `redirectURI` from the authorization request for verification.  If these values
+   * are validated, the application issues a Taskcluster token on behalf of the user who
+   * authorized the code.
+   */
+  server.exchange(oauth2orize.exchange.code(async (client, code, redirectURI, done) => {
+    const entry = await AuthorizationCode.load({ code }, true);
+
+    if (!entry) {
+      return done(null, false);
+    }
+
+    if (redirectURI !== entry.redirectUri) {
+      return done(null, false);
+    }
+
+    await AccessToken.create({
+      // OAuth2 client
+      clientId: entry.clientId,
+      redirectUri: redirectURI,
+      identity: entry.identity,
+      identityProviderId: entry.identityProviderId,
+      accessToken: entry.accessToken,
+      // This table is used alongside the AuthorizationCode table which has a 10 minute recommended expiration
+      expires: taskcluster.fromNow('10 minutes'),
+      clientDetails: entry.clientDetails,
+    }, true);
+
+    return done(null, entry.accessToken);
+  }));
+
+  const authorization = [
+    ensureLoggedIn,
+    server.authorization((clientID, redirectURI, scope, done) => {
+      const client = findRegisteredClient(clientID);
+
+      if (!client) {
+        return done(null, false);
+      }
+
+      if (!client.redirectUri.some(uri => uri === redirectURI)) {
+        return done(null, false);
+      }
+
+      return done(null, client, redirectURI);
+    }, async (client, user, scope, done) => {
```

Comment:
This `immediate` argument appears undocumented in oauth2orize (but most of the library is undocumented).  As I understand it, it's called to decide whether to immediately approve the request and return a redirect to the `redirect_uri`.  That's the second argument to `done` here.  Since it's undocumented, a comment explaining that might help the next reviewer :)